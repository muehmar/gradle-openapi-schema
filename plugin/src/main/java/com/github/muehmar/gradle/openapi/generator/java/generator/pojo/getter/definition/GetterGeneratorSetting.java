package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.definition;

public enum GetterGeneratorSetting {
  NO_JAVA_DOC,
  NO_VALIDATION,
  NO_JSON,
  PACKAGE_PRIVATE
}
