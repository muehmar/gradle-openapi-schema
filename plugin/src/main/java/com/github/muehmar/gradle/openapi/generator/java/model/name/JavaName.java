package com.github.muehmar.gradle.openapi.generator.java.model.name;

import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import lombok.EqualsAndHashCode;

/**
 * Name for Java identifiers, like member, method or classnames. A {@link JavaName} can be
 * constructed from any string and gets escaped to conform to the specification for identifiers.
 * Render the name as string with {@link JavaName#asString()} or {@link JavaName#toString()} will
 * result in the escaped string.<br>
 * <br>
 * Any operation like {@link JavaName#prefix(String)} is performed on the original string rather
 * than on the escaped string. This will produce identifiers which are as close as possible to the
 * original name, i.e. only do escaping if necessary. E.g. if a Java keyword is used as name and
 * gets prefixed, there is no longer the need to escape it.
 */
@EqualsAndHashCode
public class JavaName {
  private final Name originalName;
  private final JavaIdentifier identifier;

  private JavaName(Name name) {
    this.originalName = name;
    this.identifier = JavaIdentifier.fromName(name);
  }

  public static JavaName fromName(Name name) {
    return new JavaName(name);
  }

  public static JavaName fromString(String name) {
    return new JavaName(Name.ofString(name));
  }

  public JavaName prefix(String prefix) {
    return new JavaName(originalName.prefix(prefix));
  }

  public JavaName append(String append) {
    return new JavaName(originalName.append(append));
  }

  public JavaName startUpperCase() {
    return new JavaName(originalName.startUpperCase());
  }

  public JavaName startLowerCase() {
    return new JavaName(originalName.startLowerCase());
  }

  public JavaName prefixedMethodName(String prefix) {
    if (prefix.isEmpty()) {
      return startLowerCase();
    } else {
      return startUpperCase().prefix(prefix);
    }
  }

  public String wordBoundaryPattern() {
    return String.format("\\b%s\\b", identifier);
  }

  public Name getOriginalName() {
    return originalName;
  }

  public String asString() {
    return identifier.asString();
  }

  @Override
  public String toString() {
    return asString();
  }
}
