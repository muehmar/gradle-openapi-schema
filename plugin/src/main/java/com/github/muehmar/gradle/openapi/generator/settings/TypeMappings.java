package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import lombok.Value;

@Value
public class TypeMappings {
  PList<ClassTypeMapping> classTypeMappings;
  PList<FormatTypeMapping> formatTypeMappings;
  PList<DtoMapping> dtoMappings;

  public static TypeMappings empty() {
    return new TypeMappings(PList.empty(), PList.empty(), PList.empty());
  }

  public static TypeMappings ofSingleClassTypeMapping(ClassTypeMapping classTypeMapping) {
    return new TypeMappings(PList.single(classTypeMapping), PList.empty(), PList.empty());
  }

  public static TypeMappings ofClassTypeMappings(ClassTypeMapping... classTypeMapping) {
    return new TypeMappings(PList.fromArray(classTypeMapping), PList.empty(), PList.empty());
  }

  public static TypeMappings ofSingleFormatTypeMapping(FormatTypeMapping formatTypeMapping) {
    return new TypeMappings(PList.empty(), PList.single(formatTypeMapping), PList.empty());
  }

  public static TypeMappings ofSingleDtoMapping(DtoMapping dtoMapping) {
    return new TypeMappings(PList.empty(), PList.empty(), PList.single(dtoMapping));
  }
}
