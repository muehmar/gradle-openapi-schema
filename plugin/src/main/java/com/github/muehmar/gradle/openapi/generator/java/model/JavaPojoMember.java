package com.github.muehmar.gradle.openapi.generator.java.model;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator.EnumContent;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsNullFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsPresentFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoMember;
import com.github.muehmar.gradle.openapi.generator.model.type.EnumType;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import lombok.With;

/**
 * This member corresponds to a property in the specification. To support all possible states af a
 * member, a single {@link JavaPojoMember} may result in more than one {@link TechnicalPojoMember}
 * in a pojo.
 */
@EqualsAndHashCode
@ToString
@PojoBuilder
@With
public class JavaPojoMember {
  private final JavaName name;
  private final String description;
  private final JavaType javaType;
  private final Necessity necessity;
  private final Nullability nullability;
  private final MemberType type;

  private static final String TRISTATE_TO_PROPERTY =
      "onValue(val -> val).onNull(() -> null).onAbsent(() -> null)";
  private static final String TRISTATE_TO_ISNULL_FLAG =
      "onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false)";

  JavaPojoMember(
      JavaName name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability,
      MemberType type) {
    this.javaType = javaType;
    this.name = name;
    this.description = description;
    this.necessity = necessity;
    this.nullability = nullability;
    this.type = type;
  }

  public static JavaPojoMember wrap(PojoMember pojoMember, TypeMappings typeMappings) {
    final JavaType javaType = JavaType.wrap(pojoMember.getType(), typeMappings);
    return new JavaPojoMember(
        JavaName.fromName(pojoMember.getName()),
        pojoMember.getDescription(),
        javaType,
        pojoMember.getNecessity(),
        pojoMember.getNullability(),
        MemberType.OBJECT_MEMBER);
  }

  public JavaName getName() {
    return name;
  }

  public Nullability getNullability() {
    return nullability;
  }

  public Necessity getNecessity() {
    return necessity;
  }

  public String getDescription() {
    return description;
  }

  public JavaType getJavaType() {
    return javaType;
  }

  public MemberType getType() {
    return type;
  }

  public JavaPojoMember asObjectMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.OBJECT_MEMBER)
        .andAllOptionals()
        .build();
  }

  public JavaPojoMember asAllOfMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.ALL_OF_MEMBER)
        .andAllOptionals()
        .build();
  }

  public JavaPojoMember asOneOfMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.ONE_OF_MEMBER)
        .andAllOptionals()
        .build();
  }

  public JavaPojoMember asAnyOfMember() {
    return JavaPojoMemberBuilder.create()
        .name(name)
        .description(description)
        .javaType(javaType)
        .necessity(necessity)
        .nullability(nullability)
        .type(MemberType.ANY_OF_MEMBER)
        .andAllOptionals()
        .build();
  }

  public boolean isOptional() {
    return necessity.isOptional();
  }

  public boolean isRequired() {
    return necessity.isRequired();
  }

  public boolean isNullable() {
    return nullability.isNullable();
  }

  public boolean isNotNullable() {
    return nullability.isNotNullable();
  }

  public boolean isRequiredAndNullable() {
    return isRequired() && isNullable();
  }

  public boolean isRequiredAndNotNullable() {
    return isRequired() && isNotNullable();
  }

  public boolean isOptionalAndNullable() {
    return isOptional() && isNullable();
  }

  public boolean isOptionalAndNotNullable() {
    return isOptional() && isNotNullable();
  }

  public JavaName getWitherName() {
    return prefixedMethodName("with");
  }

  public JavaName getIsPresentFlagName() {
    return IsPresentFlagName.fromName(name).getName();
  }

  public JavaName getIsNullFlagName() {
    return IsNullFlagName.fromName(name).getName();
  }

  public JavaName getGetterName() {
    return prefixedMethodName("get");
  }

  public JavaName getValidationGetterName(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString())
        .append(settings.getValidationMethods().getGetterSuffix());
  }

  public JavaName getGetterNameWithSuffix(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString()).append(determineSuffix(settings));
  }

  private String determineSuffix(PojoSettings settings) {
    final GetterSuffixes getterSuffixes = settings.getGetterSuffixes();
    if (isRequiredAndNotNullable()) {
      return getterSuffixes.getRequiredSuffix();
    } else if (isRequiredAndNullable()) {
      return getterSuffixes.getRequiredNullableSuffix();
    } else if (isOptionalAndNotNullable()) {
      return getterSuffixes.getOptionalSuffix();
    } else {
      return getterSuffixes.getOptionalNullableSuffix();
    }
  }

  public JavaName prefixedMethodName(String prefix) {
    return name.prefixedMethodName(prefix);
  }

  public PList<TechnicalPojoMember> getTechnicalMembers() {
    final TechnicalPojoMember technicalPojoMember = TechnicalPojoMember.wrapPojoMember(this);
    if (isRequiredAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isPresentFlagMember(getIsPresentFlagName()));
    } else if (isOptionalAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isNullFlagMember(getIsNullFlagName()));
    } else {
      return PList.single(technicalPojoMember);
    }
  }

  public String tristateToProperty() {
    return TRISTATE_TO_PROPERTY;
  }

  public String tristateToIsNullFlag() {
    return TRISTATE_TO_ISNULL_FLAG;
  }

  /** Creates {@link EnumContent} for this member in case its type is an {@link EnumType}. */
  public Optional<EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumContent>> toEnumPojo =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaName.fromName(enumType.getQualifiedClassName().getClassName()))
                    .description(getDescription())
                    .members(enumType.getMembers())
                    .build());
    return javaType.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        toEnumPojo,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }

  /**
   * In case this member is an enum type, it converts the classname of the enum so that it's
   * referenced via outer-class {@code javaPojoName}.
   */
  public JavaPojoMember asInnerEnumOf(JavaName javaPojoName) {
    final JavaType newType =
        javaType.fold(
            arrayType -> arrayType,
            booleanType -> booleanType,
            enumType -> enumType.asInnerClassOf(javaPojoName),
            mapType -> mapType,
            noType -> noType,
            numericType -> numericType,
            integerType -> integerType,
            objectType -> objectType,
            stringType -> stringType);
    return new JavaPojoMember(name, description, newType, necessity, nullability, type);
  }

  public enum MemberType {
    OBJECT_MEMBER(false),
    ALL_OF_MEMBER(true),
    ONE_OF_MEMBER(true),
    ANY_OF_MEMBER(true),
    ARRAY_VALUE(false);

    private final boolean isComposedMember;

    MemberType(boolean isComposedMember) {
      this.isComposedMember = isComposedMember;
    }

    public boolean isComposedMember() {
      return isComposedMember;
    }
  }
}
