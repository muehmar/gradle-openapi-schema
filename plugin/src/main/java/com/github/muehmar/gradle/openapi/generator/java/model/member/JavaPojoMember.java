package com.github.muehmar.gradle.openapi.generator.java.model.member;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator.EnumContent;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsNotNullFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsNullFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.IsPresentFlagName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.PropertyInfoName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoMember;
import com.github.muehmar.gradle.openapi.generator.model.type.EnumType;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import lombok.With;

/**
 * This member corresponds to a property in the specification. To support all possible states af a
 * member, a single {@link JavaPojoMember} may result in more than one {@link TechnicalPojoMember}
 * in a pojo.
 */
@EqualsAndHashCode
@ToString
@PojoBuilder
@With
public class JavaPojoMember {
  private final JavaPojoName pojoName;
  private final JavaName name;
  private final String description;
  private final JavaType javaType;
  private final Necessity necessity;
  private final Nullability nullability;
  private final MemberType type;

  private static final String TRISTATE_TO_PROPERTY =
      "onValue(val -> val).onNull(() -> null).onAbsent(() -> null)";
  private static final String TRISTATE_TO_ISNULL_FLAG =
      "onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false)";

  JavaPojoMember(
      JavaPojoName pojoName,
      JavaName name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability,
      MemberType type) {
    this.pojoName = pojoName;
    this.javaType = javaType;
    this.name = name;
    this.description = description;
    this.necessity = necessity;
    this.nullability = nullability;
    this.type = type;
  }

  public static JavaPojoMember wrap(
      PojoMember pojoMember, JavaPojoName pojoName, TypeMappings typeMappings) {
    final JavaType javaType = JavaType.wrap(pojoMember.getType(), typeMappings);
    return new JavaPojoMember(
        pojoName,
        JavaName.fromName(pojoMember.getName()),
        pojoMember.getDescription(),
        javaType,
        pojoMember.getNecessity(),
        pojoMember.getNullability(),
        MemberType.OBJECT_MEMBER);
  }

  public PropertyInfoName getPropertyInfoName() {
    return PropertyInfoName.fromPojoNameAndMemberName(pojoName, name);
  }

  public JavaName getName() {
    return name;
  }

  public Nullability getNullability() {
    return nullability;
  }

  public Necessity getNecessity() {
    return necessity;
  }

  public String getDescription() {
    return description;
  }

  public JavaType getJavaType() {
    return javaType;
  }

  public MemberType getType() {
    return type;
  }

  public boolean isTechnicallyEquals(JavaPojoMember other) {
    return this.getMemberKey().equals(other.getMemberKey());
  }

  public MemberKey getMemberKey() {
    return new MemberKey(name, javaType);
  }

  public JavaPojoMember asAllOfMember() {
    return withType(MemberType.ALL_OF_MEMBER);
  }

  public JavaPojoMember asOneOfMember() {
    return withType(MemberType.ONE_OF_MEMBER);
  }

  public JavaPojoMember asAnyOfMember() {
    return withType(MemberType.ANY_OF_MEMBER);
  }

  public boolean isOptional() {
    return necessity.isOptional();
  }

  public boolean isRequired() {
    return necessity.isRequired();
  }

  public boolean isNullable() {
    return nullability.isNullable();
  }

  public boolean isNotNullable() {
    return nullability.isNotNullable();
  }

  public boolean isRequiredAndNullable() {
    return isRequired() && isNullable();
  }

  public boolean isRequiredAndNotNullable() {
    return isRequired() && isNotNullable();
  }

  public boolean isOptionalAndNullable() {
    return isOptional() && isNullable();
  }

  public boolean isOptionalAndNotNullable() {
    return isOptional() && isNotNullable();
  }

  public JavaName getWitherName() {
    return prefixedMethodName("with");
  }

  public JavaName getIsPresentFlagName() {
    return IsPresentFlagName.fromName(name).getName();
  }

  public JavaName getIsNullFlagName() {
    return IsNullFlagName.fromName(name).getName();
  }

  public JavaName getIsNotNullFlagName() {
    return IsNotNullFlagName.fromName(name).getName();
  }

  public JavaName getGetterName() {
    return prefixedMethodName("get");
  }

  public JavaName getValidationGetterName(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString())
        .append(settings.getValidationMethods().getGetterSuffix());
  }

  public JavaName getGetterNameWithSuffix(PojoSettings settings) {
    return JavaName.fromString(getGetterName().asString()).append(determineSuffix(settings));
  }

  private String determineSuffix(PojoSettings settings) {
    final GetterSuffixes getterSuffixes = settings.getGetterSuffixes();
    if (isRequiredAndNotNullable()) {
      return getterSuffixes.getRequiredSuffix();
    } else if (isRequiredAndNullable()) {
      return getterSuffixes.getRequiredNullableSuffix();
    } else if (isOptionalAndNotNullable()) {
      return getterSuffixes.getOptionalSuffix();
    } else {
      return getterSuffixes.getOptionalNullableSuffix();
    }
  }

  public JavaName prefixedMethodName(String prefix) {
    return name.prefixedMethodName(prefix);
  }

  public PList<TechnicalPojoMember> getTechnicalMembers() {
    final TechnicalPojoMember technicalPojoMember = TechnicalPojoMember.wrapPojoMember(this);
    if (isRequiredAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isPresentFlagMember(getIsPresentFlagName()));
    } else if (isOptionalAndNotNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isNotNullFlagMember(getIsNotNullFlagName()));
    } else if (isOptionalAndNullable()) {
      return PList.of(
          technicalPojoMember, TechnicalPojoMember.isNullFlagMember(getIsNullFlagName()));
    } else {
      return PList.single(technicalPojoMember);
    }
  }

  public String tristateToProperty() {
    return TRISTATE_TO_PROPERTY;
  }

  public String tristateToIsNullFlag() {
    return TRISTATE_TO_ISNULL_FLAG;
  }

  public Optional<JavaPojoMember> mergeToLeastRestrictive(JavaPojoMember other) {
    if (this.getMemberKey().equals(other.getMemberKey())) {
      final Nullability leastRestrictiveNullability =
          Nullability.leastRestrictive(this.getNullability(), other.getNullability());
      final Necessity leastRestrictiveNecessity =
          Necessity.leastRestrictive(this.getNecessity(), other.getNecessity());
      return Optional.of(
          this.withNullability(leastRestrictiveNullability)
              .withNecessity(leastRestrictiveNecessity));
    } else {
      return Optional.empty();
    }
  }

  public Optional<JavaPojoMember> mergeToMostRestrictive(JavaPojoMember other) {
    if (this.getMemberKey().equals(other.getMemberKey())) {
      final Nullability mostRestrictiveNullability =
          Nullability.mostRestrictive(this.getNullability(), other.getNullability());
      final Necessity mostRestrictiveNecessity =
          Necessity.mostRestrictive(this.getNecessity(), other.getNecessity());
      return Optional.of(
          this.withNullability(mostRestrictiveNullability).withNecessity(mostRestrictiveNecessity));
    } else {
      return Optional.empty();
    }
  }

  /** Creates {@link EnumContent} for this member in case its type is an {@link EnumType}. */
  public Optional<EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumContent>> toEnumPojo =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaName.fromName(enumType.getQualifiedClassName().getClassName()))
                    .description(getDescription())
                    .members(enumType.getMembers())
                    .build());
    return javaType.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        toEnumPojo,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }

  /**
   * In case this member is an enum type, it converts the classname of the enum so that it's
   * referenced via outer-class {@code javaPojoName}.
   */
  public JavaPojoMember asInnerEnumOf(JavaName javaPojoName) {
    final JavaType newType =
        javaType.fold(
            arrayType -> arrayType,
            booleanType -> booleanType,
            enumType -> enumType.asInnerClassOf(javaPojoName),
            mapType -> mapType,
            noType -> noType,
            numericType -> numericType,
            integerType -> integerType,
            objectType -> objectType,
            stringType -> stringType);
    return new JavaPojoMember(pojoName, name, description, newType, necessity, nullability, type);
  }

  public enum MemberType {
    OBJECT_MEMBER,
    ADDITIONAL_PROPERTY_MEMBER,
    ALL_OF_MEMBER,
    ONE_OF_MEMBER,
    ANY_OF_MEMBER,
    ARRAY_VALUE
  }
}
