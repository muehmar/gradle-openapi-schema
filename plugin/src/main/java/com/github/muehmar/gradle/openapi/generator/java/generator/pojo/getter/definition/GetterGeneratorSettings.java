package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.definition;

import static com.github.muehmar.gradle.openapi.generator.java.GeneratorUtil.noSettingsGen;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator.jsonIgnore;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator.jsonProperty;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.validation.ValidationAnnotationGenerator.validationAnnotationsForMember;
import static com.github.muehmar.gradle.openapi.util.Booleans.not;
import static io.github.muehmar.codegenerator.java.JavaDocGenerator.javaDoc;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import java.util.function.Predicate;
import lombok.Value;

@Value
public class GetterGeneratorSettings {
  PList<GetterGeneratorSetting> settings;

  public static GetterGeneratorSettings empty() {
    return new GetterGeneratorSettings(PList.empty());
  }

  public static GetterGeneratorSettings getterGeneratorSettings(
      GetterGeneratorSetting... settings) {
    return new GetterGeneratorSettings(PList.fromArray(settings));
  }

  public <T> Predicate<T> validationFilter() {
    return ignore -> isValidation();
  }

  public boolean isValidation() {
    return settings.filter(GetterGeneratorSetting.NO_VALIDATION::equals).isEmpty();
  }

  public <T> Predicate<T> jsonFilter() {
    return ignore -> isJson();
  }

  public boolean isJson() {
    return settings.filter(GetterGeneratorSetting.NO_JSON::equals).isEmpty();
  }

  public boolean isPackagePrivate() {
    return settings.filter(GetterGeneratorSetting.PACKAGE_PRIVATE::equals).nonEmpty();
  }

  public boolean isJavaDoc() {
    return settings.filter(GetterGeneratorSetting.NO_JAVA_DOC::equals).isEmpty();
  }

  public Generator<JavaPojoMember, PojoSettings> jsonIgnoreGenerator() {
    return Generator.<JavaPojoMember, PojoSettings>emptyGen()
        .append(jsonIgnore())
        .filter(ignore -> not(this.isJson()));
  }

  public Generator<JavaPojoMember, PojoSettings> jsonPropertyGenerator() {
    return Generator.<JavaPojoMember, PojoSettings>emptyGen()
        .append(jsonProperty())
        .filter(ignore -> this.isJson());
  }

  public Generator<JavaPojoMember, PojoSettings> javaDocGenerator() {
    return Generator.<JavaPojoMember, PojoSettings>emptyGen()
        .append(noSettingsGen(javaDoc()), JavaPojoMember::getDescription)
        .filter(ignore -> this.isJavaDoc());
  }

  public Generator<JavaPojoMember, PojoSettings> validationAnnotationGenerator() {
    return Generator.<JavaPojoMember, PojoSettings>emptyGen()
        .append(validationAnnotationsForMember())
        .filter(validationFilter());
  }
}
