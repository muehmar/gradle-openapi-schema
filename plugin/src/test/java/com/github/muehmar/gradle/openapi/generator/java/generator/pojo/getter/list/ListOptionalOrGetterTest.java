package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.list;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.definition.GetterGeneratorSetting.NO_JAVA_DOC;
import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.list.ListOptionalOrGetter.listOptionalOrGetterGenerator;
import static com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers.optionalListWithNullableItems;
import static com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers.optionalStringList;
import static com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers.requiredNullableListWithNullableItems;
import static com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers.requiredNullableStringList;
import static com.github.muehmar.gradle.openapi.generator.settings.ClassTypeMappings.LIST_MAPPING_WITH_CONVERSION;
import static com.github.muehmar.gradle.openapi.generator.settings.ClassTypeMappings.STRING_MAPPING_WITH_CONVERSION;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.definition.GetterGeneratorSetting;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.definition.GetterGeneratorSettings;
import com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

@SnapshotTest
class ListOptionalOrGetterTest {
  private Expect expect;

  @ParameterizedTest
  @MethodSource("listMembers")
  @SnapshotName("listMembers")
  void generate_when_listMembers_then_matchSnapshot(JavaPojoMember member) {
    final Generator<JavaPojoMember, PojoSettings> generator =
        listOptionalOrGetterGenerator(GetterGeneratorSettings.empty());

    final Writer writer = generator.generate(member, defaultTestSettings(), javaWriter());

    expect.scenario(member.getName().asString()).toMatchSnapshot(writerSnapshot(writer));
  }

  private static Stream<Arguments> listMembers() {
    final TypeMappings fullTypeMappings =
        TypeMappings.ofClassTypeMappings(
            STRING_MAPPING_WITH_CONVERSION, LIST_MAPPING_WITH_CONVERSION);
    return Stream.of(
            requiredNullableStringList(),
            requiredNullableListWithNullableItems(),
            optionalListWithNullableItems(fullTypeMappings)
                .withName(JavaName.fromString("optionalListWithNullableItemsFullMapping")))
        .map(Arguments::arguments);
  }

  @ParameterizedTest
  @MethodSource("generatorSettings")
  @SnapshotName("generatorSettings")
  void generate_when_generatorSettings_then_matchSnapshot(
      GetterGeneratorSettings generatorSettings) {
    final Generator<JavaPojoMember, PojoSettings> generator =
        listOptionalOrGetterGenerator(generatorSettings);

    final Writer writer;
    writer = generator.generate(optionalStringList(), defaultTestSettings(), javaWriter());

    expect
        .scenario(generatorSettings.getSettings().mkString("|"))
        .toMatchSnapshot(writerSnapshot(writer));
  }

  private static Stream<Arguments> generatorSettings() {
    return Stream.<PList<GetterGeneratorSetting>>of(PList.empty(), PList.of(NO_JAVA_DOC))
        .map(GetterGeneratorSettings::new)
        .map(Arguments::arguments);
  }
}
