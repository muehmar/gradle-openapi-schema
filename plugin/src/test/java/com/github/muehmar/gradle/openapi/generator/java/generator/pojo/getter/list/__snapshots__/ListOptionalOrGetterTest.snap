generatorSettings[NO_JAVA_DOC]=[
com.fasterxml.jackson.annotation.JsonIgnore
java.util.Optional
java.util.function.Function

@JsonIgnore
public List<String> getOptionalStringListOr(List<String> defaultValue) {
  return this.optionalStringList == null
      ? defaultValue
      : mapList(
          optionalStringList,
          Function.identity(),
          Function.identity(),
          Function.identity(),
          Function.identity()
      );
}
]


generatorSettings[]=[
com.fasterxml.jackson.annotation.JsonIgnore
java.util.Optional
java.util.function.Function

/**
 * List
 */
@JsonIgnore
public List<String> getOptionalStringListOr(List<String> defaultValue) {
  return this.optionalStringList == null
      ? defaultValue
      : mapList(
          optionalStringList,
          Function.identity(),
          Function.identity(),
          Function.identity(),
          Function.identity()
      );
}
]


listMembers[optionalListWithNullableItemsFullMapping]=[
com.custom.CustomList
com.custom.CustomString
com.fasterxml.jackson.annotation.JsonIgnore
java.util.Optional
java.util.function.Function

/**
 * OptionalListWithNullableItems
 */
@JsonIgnore
public CustomList<Optional<CustomString>> getOptionalListWithNullableItemsFullMappingOr(CustomList<Optional<CustomString>> defaultValue) {
  return this.optionalListWithNullableItemsFullMapping == null
      ? defaultValue
      : mapList(
          optionalListWithNullableItemsFullMapping,
          i -> CustomString.fromString(i),
          Optional::ofNullable,
          l -> CustomList.fromList(l),
          Function.identity()
      );
}
]


listMembers[requiredNullableListWithNullableItems]=[
com.fasterxml.jackson.annotation.JsonIgnore
java.util.Optional
java.util.function.Function

/**
 * RequiredNullableListWithNullableItems
 */
@JsonIgnore
public List<Optional<String>> getRequiredNullableListWithNullableItemsOr(List<Optional<String>> defaultValue) {
  return this.requiredNullableListWithNullableItems == null
      ? defaultValue
      : mapList(
          requiredNullableListWithNullableItems,
          Function.identity(),
          Optional::ofNullable,
          Function.identity(),
          Function.identity()
      );
}
]


listMembers[requiredNullableStringList]=[
com.fasterxml.jackson.annotation.JsonIgnore
java.util.Optional
java.util.function.Function

/**
 * List
 */
@JsonIgnore
public List<String> getRequiredNullableStringListOr(List<String> defaultValue) {
  return this.requiredNullableStringList == null
      ? defaultValue
      : mapList(
          requiredNullableStringList,
          Function.identity(),
          Function.identity(),
          Function.identity(),
          Function.identity()
      );
}
]